### ISSUE TITLE
PSM3 issues

### ISSUE THREAD

**Issue by jsquyres** [ofiwg/libfabric#6661]
From https://github.com/ofiwg/libfabric/pull/6631#issuecomment-804238843, here's what I see is still broken with the PSM3 provider:

* Do not re-invent Make or Automake or Autoconf mechanisms
    * Namely: the `.libs/*.o` rules in `prov/psm3/Makefile.include` are very fragile, and should probably be done with C code, not hand-editing private/undocumented `.la` files and linker tomfoolery.
* Do not hijack global rules
    * Namely: the `all-local` rule in `prov/psm3/Makefile.include`
* If you're going to only do the checksum thing once, do it during configure. Otherwise, do it properly during the build.
    * Right now, it is very broken.
* Uplevel the checksum thing.

Additionally:

* Remove redundant `configure.ac` and `Makefile.am` files (and probably the various `*.in` files that do not appear to be used by `prov/psm3/configure.m4`).  If these are not part of libfabric, they should not be in this repo / tree.
* The `README` file does not look appropriate for libfabric, either.

Bottom line: the libfabric repo is not an appropriate dumping ground for all psm3 development.  If you have separate, non-libfabric psm3 products, please maintain that elsewhere.  Thanks!

You may already be working on fixing these things; just filing them here so that they're tracked properly.

**Comment by acgoldma** [comment]
The other configure/makefiles in the repo are to build an Out-of-tree provider that can work with older versions of libfabric (in-dsitro). Current DL providers only work with the libfabric they are built with. 
Either way, since they are tracked in our (Intel's) release github ([intel/eth-psm3-fi](https://github.com/intel/eth-psm3-fi)), I have no issue removing them from libfabric repo. Same with README.

As for the others, the checksum is being worked on.
As for the first bullet (the 're-invention' of automake mechanisms), we are also discussing internally, as we are trying to figure out the best way to maintain some internal testing abilities.

Thank you for the help.

**Comment by shefty** [comment]
I don't have an issue keeping the files needed to build an DL provider upstream.  I think that's reasonable, much like maintaining the vendor CI scripts.

Someday we can hopefully make the DL support more generic, so it's easier to add that support for all providers.

**Comment by acgoldma** [comment]
Did not want this to close, just link. Here is a link to Checksum/all-local bullets (2 and 3). #6681

**Comment by jsquyres** [comment]
@acgoldma and I had a good chat today on webex and sorted out most of the issues (e.g., as he cited above, #6681 solves the checksum issues -- it would still be great to get that upleveled; @shefty can this be added to a Tuesday agenda next week? #6681 is a good blueprint for how it could be done at the core level).

The last issue appears to be the rules at the end of `prov/psm3/Makefile.include`.  Adam explained to me what they are for: to hide the fact that PSM2 and PSM3 share a lot of the same symbol names.

This seems like an implementation choice that shouldn't be part of libfabric.  Specifically: providers should not knowingly use the same symbols as another provider.  Meaning: please choose different symbol names so that you do not clash with other providers.  This will allow deleting of those fragile rules in `prov/psm3/Makefile.include`, and thereby stop using (and re-implementing) non-public, undocumented, and internal Libtool mechanisms that are subject to change at any time.

**Comment by shefty** [comment]
Yes, I can add discussion in the ofiwg of whether it's worthwhile to add a checksum feature at the top of libfabric.

And I agree that psm3 should use their own prefix names.  My understanding is that the symbol clash is a result of the upstream psm3 provider code with out of tree psm2 source files.

**Comment by acgoldma** [comment]
> And I agree that psm3 should use their own prefix names. My understanding is that the symbol clash is a result of the upstream psm3 provider code with out of tree psm2 source files.

Correct, it is only when a user also uses the `--with-psm2-src` configure flag which embeds libpsm2 src into libfabric that the internal symbols conflict with psm3 provider.
